USER_GUIDE.md

```diff
@@ -213,6 +213,69 @@ do {
 
 Note that we are passing the Jar file in the `classPath` argument when initializing the `JavaVirtualMachine` instance. Otherwise, the program will fail with an error because it cannot find the Java class `com.gazman.quadratic_sieve.primes.SieveOfEratosthenes`.
 
+### Up and downcasting
+
+All Java classes available in Swift provide `is` and `as` methods to check whether an object dynamically matches another type. The `is` operation is the equivalent of Java's `instanceof` and Swift's `is` operator, and will checkin whether a given object is of the specified type, e.g.,
+
+```swift
+if myObject.is(URL.self) {
+  // myObject is a Java URL.
+}
+```
+
+Often, one also wants to cast to that type. The `as` method returns an optional of the specified type, so it works well with `if let`:
+
+```swift
+if let url = myObject.as(URL.self) {
+  // okay, url is a Java URL
+}
+```
+
+> *Note*: The Swift `is`, `as?`, and `as!` operators do *not* work correctly with the Swift projections of Java types. Use the `is` and `as` methods consistently.
+
+### Implementing Java `native` methods in Swift
+
+JavaKit supports implementing Java `native` methods in Swift using JNI with the `@JavaImplementation` macro. In Java, the method must be declared as `native`, e.g.,
+
+```java
+package org.swift.javakit.example;
+
+public class HelloSwift {
+    static {
+        System.loadLibrary("HelloSwiftLib");
+    }
+
+    public native String reportStatistics(String meaning, double[] numbers);
+}
+```
+
+On the Swift side, the Java class needs to be exposed to Swift through `Java2Swift.config`, e.g.,:
+
+```swift
+{
+  "classes" : {
+    "org.swift.javakit.example.HelloSwift" : "Hello",
+  }
+}
+```
+
+Implementations of `native` methods are written in an extension of the Swift type that has been marked with `@JavaImplementation`. The methods themselves must be marked with `@JavaMethod`, indicating that they are available to Java as well. To help ensure that the Swift code implements all of the `native` methods with the right signatures, JavaKit produces a protocol with the Swift type name suffixed by `NativeMethods`. Declare conformance to that protocol and implement its requirements, for example:
+
+```swift
+@JavaImplementation("org.swift.javakit.HelloSwift")
```

Comment:
Just a thought on the names matching dance here: can we do better than that? This currently needs to match up the java file, the config file, and the swift file. Can the config file eventually become an implementation detail? It seems like a moving piece that folks are bound to forget about when e.g. moving the java type to some other package 🤔 

Perhaps just more good diagnostics would be a way to address this as well.

Occurred to me now while reviewing and comparing with how much of a hassle this is vs integrating Java/JNI/C 🤔

Reply:
I think it depends on how much work we're willing to do in the plugin itself. The config file lets us predetermine all of the `.swift` files that will be generated by reading just the one file. If we want to extract that information from the Swift or Java sources, we have to do a lot more digging just to set up the build graph.

The *NativeMethods protocol should help a bit with the names-matching, though. You write the extension to conform to protocol, and the compiler makes sure that all of the names and types match the `native` function. 

Good diagnostics would definitely help here, and we should think more about whether it's okay to generate more of the config file to save steps.